RandLAPACK/comps/rl_util.hh

```diff
@@ -370,6 +370,189 @@ void gen_exp_mat(
     }
 }
 
+/// Generates matrix with a staircase spectrum with 4 steps
+template <typename T>
+void gen_step_mat(
+    int64_t& m,
+    int64_t& n,
+    std::vector<T>& A,
+    int64_t k,
+    T cond,
+    bool diagon,
+    RandBLAS::base::RNGState<r123::Philox4x32> state
+) {
+
+    // Predeclare to all nonzero constants, start decay where needed
+    std::vector<T> s(k, 1.0);
+    std::vector<T> S(k * k, 0.0);
+
+    // We will have 4 steps controlled by the condition number size and starting with 1
+    int offset = (int) (k / 4);
+
+    std::fill(s.begin(), s.begin() + offset, 1);
+    std::fill(s.begin() + offset + 1, s.begin() + 2 * offset, 8.0 / cond);
+    std::fill(s.begin() + 2 * offset + 1, s.begin() + 3 * offset, 4.0 / cond);
+    std::fill(s.begin() + 3 * offset + 1, s.end(), 1.0 / cond);
+
+    // form a diagonal S
+    diag(k, k, s, k, S);
+
+    if (diagon) {
+        if (!(m == k || n == k)) {
+            m = k;
+            n = k;
+            A.resize(k * k);
+        }
+        lapack::lacpy(MatrixType::General, k, k, S.data(), k, A.data(), k);
+    } else {
+        gen_mat(m, n, A, k, S, state);
+    }
+}
+
+/// Generates a matrix with high coherence between the left singular vectors.
+/// Such matrix would be difficult to sketch.
+template <typename T>
+void gen_spiked_mat(
+    int64_t& m,
+    int64_t& n,
+    std::vector<T>& A,
+    RandBLAS::base::RNGState<r123::Philox4x32> state
+) {
+
+    T* A_dat = A.data();
+
+    int64_t num_rows_sampled = n / 2;
+
+    RandBLAS::sparse::SparseDist DS = {RandBLAS::sparse::SparseDistName::LASO, num_rows_sampled, m, 1};
+    RandBLAS::sparse::SparseSkOp<T> S(DS, state, NULL, NULL, NULL);
+    RandBLAS::sparse::fill_sparse(S);
+
+    // Fill A with stacked identities
+    int start = 0;
+    while(start + n <= m){
+        for(int j = 0; j < n; ++j) {
+            A_dat[start + (m * j) + j] = 1.0;
+        }
+        start += n;
+    }
+
+    // Scale randomly sampled rows
+    start = 0;
+    while (start + m <= m * n) {
+        for(int i = 0; i < num_rows_sampled; ++i) {
+            A_dat[start + (S.cols)[i] - 1] *= m;
+        }
+        start += m;
+    }
+
+    std::vector<T> A_hat(m * n, 0.0);
+    std::vector<T> V(n * n, 0.0);
+    std::vector<T> tau(n, 0.0);
+
+    RandBLAS::dense::DenseDist DV{.n_rows = n, .n_cols = n};
+    state = RandBLAS::dense::fill_buff(V.data(), DV, state);
+
+    lapack::geqrf(n, n, V.data(), n, tau.data());
+    lapack::ungqr(n, n, n, V.data(), n, tau.data());
+
+    std::copy(A.data(), A.data() + (m * n), A_hat.data());
+    blas::gemm(Layout::ColMajor, Op::NoTrans, Op::Trans, m, n, n, 1.0, A_hat.data(), m, V.data(), n, 0.0, A.data(), m);
+}
+
+/// Generates a numerically rank-deficient matrix.
+/// Added per Oleg's suggestion.
+template <typename T>
+void gen_scaled_mat(
```

Comment:
I suggest we use a more informative function name.

Perhaps ``gen_oleg_adversarial_mat``. (This would not be a great name but it would at least remind us to revisit the naming question later on.)

Reply:
Updated it.